staging/src/k8s.io/pod-security-admission/cmd/webhook/app/server.go

```diff
@@ -0,0 +1,261 @@
+		klog.ErrorS(err, "unable to decode the request")
+		writeAdmissionResponse(w, admission.BadRequestResponse("unable to decode the request"))
+		return
+	}
+	klog.V(1).InfoS("received request", "UID", review.Request.UID, "kind", review.Request.Kind, "resource", review.Request.Resource)
+
+	attributes := admission.RequestAttributes(review.Request, s)
+	response := s.delegate.Validate(ctx, attributes)
+	review.Response = response
+	writeResponse(w, review)
+}
+
+// Server implements the admission.Decoder interface.
+func (s *Server) DecodeRaw(rawObj runtime.RawExtension, into runtime.Object) error {
+	if len(rawObj.Raw) == 0 {
+		return fmt.Errorf("empty object")
+	}
+	deserializer := codecs.UniversalDeserializer()
+	return runtime.DecodeInto(deserializer, rawObj.Raw, into)
+}
+
+// Setup creates an Admission object to handle the admission logic.
+func Setup(opts *options.Options) (*Server, error) {
+	if errs := opts.Validate(); len(errs) > 0 {
+		return nil, utilerrors.NewAggregate(errs)
+	}
+
+	s := &Server{}
+	opts.SecureServing.ApplyTo(&s.secureServing)
+	opts.InsecureServing.ApplyTo(&s.insecureServing)
+
+	if s.secureServing == nil && s.insecureServing == nil {
+		return nil, errors.New("no serving options configured")
+	}
+
+	// Load Kube Client
+	kubeConfig, err := clientcmd.BuildConfigFromFlags("", opts.KubeConfigFile)
+	if err != nil {
+		return nil, err
+	}
+	restclient.AddUserAgent(kubeConfig, "podsecurity-webhook")
+	client, err := clientset.NewForConfig(kubeConfig)
+	if err != nil {
+		return nil, err
+	}
+	s.informerFactory = kubeinformers.NewSharedInformerFactory(client, time.Second*30)
+	namespaceInformer := s.informerFactory.Core().V1().Namespaces()
+	namespaceLister := namespaceInformer.Lister()
+	podLister := s.informerFactory.Core().V1().Pods().Lister() // FIXME: consider switching to just use the client
+
+	// Load PodSecurity config
+	config, err := podsecurityconfigloader.LoadFromFile(opts.ConfigFile)
+	if err != nil {
+		return nil, err
+	}
+
+	evaluator, err := policy.NewEvaluator(policy.DefaultChecks())
+	if err != nil {
+		return nil, fmt.Errorf("could not create PodSecurityRegistry: %w", err)
+	}
+
+	s.delegate = &admission.Admission{
+		Configuration:    config,
+		Evaluator:        evaluator,
+		Metrics:          nil, // TODO: wire to default prometheus metrics
+		PodSpecExtractor: admission.DefaultPodSpecExtractor{},
+		PodLister:        admission.PodListerFromInformer(podLister),
+		NamespaceGetter:  admission.NamespaceGetterFromListerAndClient(namespaceLister, client),
+	}
+
+	if err := s.delegate.CompleteConfiguration(); err != nil {
+		return nil, fmt.Errorf("configuration error: %w", err)
+	}
+	if err := s.delegate.ValidateConfiguration(); err != nil {
+		return nil, fmt.Errorf("invalid configuration: %w", err)
+	}
+
+	return s, nil
+}
+
+func writeAdmissionResponse(w http.ResponseWriter, response *admissionv1.AdmissionResponse) {
```

Comment:
something needs to set response.UID to match request.UID or the API server won't accept the response as valid... where is that being done?